src/System.Net.Http/src/System/Net/Http/Managed/AuthenticateAndRedirectHandler.cs

```diff
@@ -0,0 +1,223 @@
+                    break;
+                }
+
+                redirectCount++;
+                if (redirectCount > _maxAutomaticRedirections)
+                {
+                    throw new HttpRequestException(SR.net_http_max_redirects);
+                }
+
+                // Set up for the automatic redirect
+                request.RequestUri = location;
+
+                if (RequestRequiresForceGet(response.StatusCode, request.Method))
+                {
+                    request.Method = HttpMethod.Get;
+                    request.Content = null;
+                }
+
+                // Do the redirect.
+                response.Dispose();
+            }
+
+            return response;
+        }
+
+        private bool RequestNeedsRedirect(HttpResponseMessage response)
+        {
+            // Return if redirect not allowed.
+            if (!_allowRedirect)
+                return false;
+
+            bool needRedirect = false;
+            switch (response.StatusCode)
+            {
+                case HttpStatusCode.Moved:
+                case HttpStatusCode.Found:
+                case HttpStatusCode.SeeOther:
+                case HttpStatusCode.TemporaryRedirect:
+                    needRedirect = true;
+                    break;
+
+                case HttpStatusCode.MultipleChoices:
+                    needRedirect = response.Headers.Location != null; // Don't redirect if no Location specified
+                    break;
+            }
+
+            return needRedirect;
+        }
+
+        private static bool RequestRequiresForceGet(HttpStatusCode statusCode, HttpMethod requestMethod)
+        {
+            if (statusCode == HttpStatusCode.Moved ||
+                statusCode == HttpStatusCode.Found ||
+                statusCode == HttpStatusCode.SeeOther ||
+                statusCode == HttpStatusCode.MultipleChoices)
+            {
+                return requestMethod == HttpMethod.Post;
+            }
+
+            return false;
+        }
+
+        private async Task<HttpResponseMessage> SendRequestAsync(HttpRequestMessage request, bool useCredentialCache, CancellationToken cancellationToken)
+        {
+            if (_credentials != null && _preAuthenticate)
+            {
+                // Try using previous digest response WWWAuthenticate header
+                if (_digestResponse != null)
+                {
+                    await AuthenticationHelper.TrySetDigestAuthToken(request, _credentials, _digestResponse, HttpKnownHeaderNames.Authorization).ConfigureAwait(false);
+                }
+                else
+                {
+                    AuthenticationHelper.TrySetBasicAuthToken(request, _credentials);
+                }
+            }
+
+            HttpResponseMessage response = await _innerHandler.SendAsync(request, cancellationToken).ConfigureAwait(false);
+            ICredentials currentCredential = useCredentialCache ? _credentials as CredentialCache : _credentials;
```

Comment:
useCredentialCache? Does this mean requireCredentialCache / onlyUseCredentialCache?

Reply:
use credential cache.